### ISSUE TITLE
Null semantics applies incorrectly for SqlFunction

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#16245]
Not sure if it would be applicable to more types but we generate this SQL,
Where_string_to_upper
```SQL
            AssertSql(
                @"SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE (UPPER([c].[CustomerID]) = N'ALFKI') AND UPPER([c].[CustomerID]) IS NOT NULL");
```
`[c].[CustomerID]` is never null since its PK and metadata is also correctly says that.

Null semantics is ending up with stale value.

Found out when I made totally unrelated change which made value of stale value change and generated correct SQL.

**Comment by maumar** [comment]
dupe of https://github.com/aspnet/EntityFrameworkCore/issues/16078

**Comment by smitpatel** [comment]
It does not seem like duplicate. Do you think #16078 would be fixed by doing something unrelated to null semantics?

**Comment by maumar** [comment]
@smitpatel you are right, its a bug, fix incoming

**Comment by smitpatel** [comment]
Fixed via #16266
